definitions/output/all/backfill_pages.js

```diff
@@ -0,0 +1,157 @@
+const iterations = []
+const clients = constants.clients
+
+let midMonth
+for (
+  let date = '2016-01-01'; // 2022-06-01
+  date >= '2016-01-01'; // 2016-01-01
+  date = constants.fnPastMonth(date)
+) {
+  clients.forEach((client) => {
+    iterations.push({
+      date,
+      client
+    })
+  })
+
+  if (date <= '2018-12-01') {
+    midMonth = new Date(date)
+    midMonth.setDate(15)
+
+    clients.forEach((client) => {
+      iterations.push({
+        date: midMonth.toISOString().substring(0, 10),
+        client
+      })
+    })
+  }
+}
+
+iterations.forEach((iteration, i) => {
+  operate(`backfill_pages ${iteration.date} ${iteration.client}`).tags([
+    'backfill_pages'
+  ]).dependencies([
+    i === 0 ? '' : `backfill_pages ${iterations[i - 1].date} ${iterations[i - 1].client}`
+  ]).queries(ctx => `
+DELETE FROM \`all_dev.pages_stable\`
+WHERE date = '${iteration.date}' AND client = '${iteration.client}';
+
+CREATE TEMPORARY FUNCTION GET_OTHER_CUSTOM_METRICS(
+  jsonObject JSON,
+  keys ARRAY<STRING>
+) RETURNS JSON
+LANGUAGE js AS """
+try {
+  let other_metrics = {};
+  keys.forEach(function(key) {
+    other_metrics[key.substr(1)] = JSON.parse(jsonObject[key]);
+  });
+  return other_metrics;
+} catch (e) {
+  return null;
+}
+""";
+
+CREATE TEMP FUNCTION GET_FEATURES(payload STRING)
+RETURNS ARRAY<STRUCT<feature STRING, id STRING, type STRING>> LANGUAGE js AS
+'''
+  function getFeatureNames(featureMap, featureType) {
+    try {
+      return Object.entries(featureMap).map(([key, value]) => {
+        // After Feb 2020 keys are feature IDs.
+        if (value.name) {
+          return {'feature': value.name, 'type': featureType, 'id': key};
+        }
+        // Prior to Feb 2020 keys fell back to IDs if the name was unknown.
+        if (idPattern.test(key)) {
+          return {'feature': '', 'type': featureType, 'id': key.match(idPattern)[1]};
+        }
+        // Prior to Feb 2020 keys were names by default.
+        return {'feature': key, 'type': featureType, 'id': ''};
+      });
+    } catch (e) {
+      return [];
+    }
+  }
+
+  var $ = JSON.parse(payload);
+  if (!$._blinkFeatureFirstUsed) return [];
+
+  var idPattern = new RegExp('^Feature_(\\\\d+)$');
+  return getFeatureNames($._blinkFeatureFirstUsed.Features, 'default')
+    .concat(getFeatureNames($._blinkFeatureFirstUsed.CSSFeatures, 'css'))
+    .concat(getFeatureNames($._blinkFeatureFirstUsed.AnimatedCSSFeatures, 'animated-css'));
+''';
+
+INSERT INTO \`all_dev.pages_stable\`  --${ctx.resolve('all', 'pages')}
+SELECT
+  DATE('${iteration.date}') AS date,
+  '${iteration.client}' AS client,
+  pages.url AS page,
+  TRUE AS is_root_page,
+  pages.url AS root_page,
+  crux.rank AS rank,
+  JSON_VALUE(payload, "$.testID") AS wptid,
```

Comment:
@pmeenan I see WPT ID is available at this path starting Feb 2021.
Where should I look for it earlier?

Or should we populate another value here?

Reply:
Looks like the `summary_pages` dataset tables have `wptid` as a distinct column. The `pages` dataset itself doesn't have it in the payload because it wasn't part of the HAR files at the time so you can't get it from there. Maybe join on `url` with the `summary_pages` or `runs` datasets?